ui/src/app/views/breadcrumb/component.spec.ts

```diff
@@ -0,0 +1,198 @@
+import {Component} from '@angular/core';
+import {async, ComponentFixture, fakeAsync, TestBed, tick} from '@angular/core/testing';
+import {Router, Routes} from '@angular/router';
+import {RouterTestingModule} from '@angular/router/testing';
+
+import {WorkspacesService} from 'generated';
+import {WorkspacesServiceStub, WorkspaceStubVariables} from 'testing/stubs/workspace-service-stub';
+import {WorkspaceResolver} from '../../resolvers/workspace';
+import {BreadcrumbComponent} from './component';
+
+// TODO: Would love a better way of reading the routes. This mock seems super icky.
+// Pulling in all of the app's routes is more painful.
+const routes: Routes = [
```

Comment:
@blrubenstein and @johnstill - I'd like your opinion on how I can mock the application's routes better. I've tried exporting all of the app's routes, but then using them in `RouterTestingModule` was particularly painful because we then need to add every route component as a declaration/provider/etc.

Reply:
Could you theoretically pull it in using deepCopy, and then iterate over the map changing the component? It may be a bit more computationally painful (but not much, I imagine), but will avoid needing to redefine in multiple places.

Reply:
I would do one of two (maybe three) things:

1.  Don't unit test some of these scenarios - they involve such sweeping knowledge of the whole codebase that they are basically end to end tests anyways, so instead of mocking the whole route table (and by extension the whole application), just write an end to end test using the real thing.
2. Since you control exactly what you're mocking, don't mock _all_ the real routes.  Make a route config with just the bare minimum of dummy routes you need (one for supported, one for unsupported, etc).  As a side note, if the test here is supposed to test that ALL of our _real_ routes are / aren't supported correctly, then (A) it doesn't b/c these are mocks using mock components and (B) you'll have to update this test with every change to the routing table, which isn't ideal.
3.  Minimize the contact our own code makes with the router, then test _our_ code thoroughly but ignore the router.  I.e. test `makeBreadCrumbs`, etc etc, but don't bother testing it as subscribed to the router.

I would probably do 3.  The template is super simple and not doing much logic, and ditto for `ngOnInit`.  So you can "shallow test" this component by just making an instance and running tests on the relevant methods that house our business logic.

Reply:
Interesting ... will try something like that.